File: src/vb/Samples/ApiSamples/Samples/Common/FileProperties.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text

Namespace ApiSamples.Common
	Friend Class FileProperties
		Public Shared Sub AddCustomProperties(ByVal propertySets As SolidEdgeFramework.PropertySets)
			Dim properties = DirectCast(propertySets.Item("Custom"), SolidEdgeFramework.Properties)
			Dim propertyValues = New Object() { "My text", Integer.MaxValue, 1.23, True, Date.Now }

			For Each propertyValue In propertyValues
				Dim propertyType = propertyValue.GetType()
				Dim propertyName = String.Format("My {0}", propertyType)
				Dim [property] = properties.Add(propertyName, propertyValue)

				Console.WriteLine("Added {0} - {1}.", [property].Name, propertyValue)
			Next propertyValue
		End Sub

		Public Shared Sub DeleteCustomProperties(ByVal propertySets As SolidEdgeFramework.PropertySets)
			Dim properties = DirectCast(propertySets.Item("Custom"), SolidEdgeFramework.Properties)

			' Query for custom properties that start with "My".
			Dim query = properties.OfType(Of SolidEdgeFramework.Property)().Where(Function(x) x.Name.StartsWith("My"))

			' Force ToArray() so that Delete() doesn't interfere with the enumeration.
			For Each [property] In query.ToArray()
				Dim propertyName = [property].Name
				[property].Delete()
				Console.WriteLine("Deleted {0}.", propertyName)
			Next [property]
		End Sub

		Public Shared Sub ReportAllProperties(ByVal propertySets As SolidEdgeFramework.PropertySets)
			For i As Integer = 1 To propertySets.Count
				Dim properties = propertySets.Item(i)

				Console.WriteLine("PropertSet '{0}'.", properties.Name)

				For j As Integer = 1 To properties.Count
					Dim nativePropertyType As System.Runtime.InteropServices.VarEnum = System.Runtime.InteropServices.VarEnum.VT_EMPTY
					Dim runtimePropertyType As Type = Nothing

					Dim value As Object = Nothing

                    Dim objProperty = properties.Item(j)
                    nativePropertyType = DirectCast(objProperty.Type, System.Runtime.InteropServices.VarEnum)

					' Accessing Value property may throw an exception...
					Try
                        value = objProperty.Value
					Catch ex As System.Exception
						value = ex.Message
					End Try

					If value IsNot Nothing Then
						runtimePropertyType = value.GetType()
					End If

                    Console.WriteLine(ControlChars.Tab & "{0} = '{1}' ({2} | {3}).", objProperty.Name, value, nativePropertyType, runtimePropertyType)
				Next j

				Console.WriteLine()
			Next i
		End Sub
	End Class
End Namespace

File: src/cs/Samples/ApiSamples/Samples/Common/FileProperties.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ApiSamples.Common
{
    class FileProperties
    {
        public static void AddCustomProperties(SolidEdgeFramework.PropertySets propertySets)
        {
            var properties = (SolidEdgeFramework.Properties)propertySets.Item("Custom");
            var propertyValues = new object[] { "My text", int.MaxValue, 1.23, true, DateTime.Now };

            foreach (var propertyValue in propertyValues)
            {
                var propertyType = propertyValue.GetType();
                var propertyName = String.Format("My {0}", propertyType);
                var property = properties.Add(propertyName, propertyValue);

                Console.WriteLine("Added {0} - {1}.", property.Name, propertyValue);
            }
        }

        public static void DeleteCustomProperties(SolidEdgeFramework.PropertySets propertySets)
        {
            var properties = (SolidEdgeFramework.Properties)propertySets.Item("Custom");

            // Query for custom properties that start with "My".
            var query = properties.OfType<SolidEdgeFramework.Property>().Where(x => x.Name.StartsWith("My"));

            // Force ToArray() so that Delete() doesn't interfere with the enumeration.
            foreach (var property in query.ToArray())
            {
                var propertyName = property.Name;
                property.Delete();
                Console.WriteLine("Deleted {0}.", propertyName);
            }
        }

        public static void ReportAllProperties(SolidEdgeFramework.PropertySets propertySets)
        {
            for (int i = 1; i <= propertySets.Count; i++)
            {
                var properties = propertySets.Item(i);

                Console.WriteLine("PropertSet '{0}'.", properties.Name);

                for (int j = 1; j <= properties.Count; j++)
                {
                    System.Runtime.InteropServices.VarEnum nativePropertyType = System.Runtime.InteropServices.VarEnum.VT_EMPTY;
                    Type runtimePropertyType = null;

                    object value = null;

                    var property = properties.Item(j);
                    nativePropertyType = (System.Runtime.InteropServices.VarEnum)property.Type;

                    // Accessing Value property may throw an exception...
                    try
                    {
                        value = property.get_Value();
                    }
                    catch (System.Exception ex)
                    {
                        value = ex.Message;
                    }

                    if (value != null)
                    {
                        runtimePropertyType = value.GetType();
                    }

                    Console.WriteLine("\t{0} = '{1}' ({2} | {3}).", property.Name, value, nativePropertyType, runtimePropertyType);
                }

                Console.WriteLine();
            }
        }
    }
}
